File: 王道数据结构/栈和队列/中缀表达式转后缀表达式.py
icp = {'#' : 0, '(' : 6, '*' : 4, '/' : 4, '+' : 2, '-' : 2, ')' : 1}
isp = {'#' : 0, '(' : 1, '*' : 5, '/' : 5, '+' : 3, '-' : 3, ')' : 6}
# 栈外(in coming priority)优先数、
# 栈内(in stack priority)优先数
stk = []
stk.append('#')
res = []
s = "a+b*c-d"
for i in (s+'#'):
    print(res, stk, i)
    if i not in icp:
        res.append(i)
    else:
        if icp[i] > isp[stk[-1]]:
            stk.append(i)
        elif icp[i] < isp[stk[-1]]:
            while icp[i] < isp[stk[-1]]:
                res.append(stk.pop())
            stk.append(i)
        elif icp[i] == isp[stk[-1]]:
            stk.pop()
print(''.join(res))

File: 王道数据结构/栈和队列/中缀表达式转后缀表达式.cpp
#include <stack>
// #include
#include <unordered_map>
#include <iostream>
using namespace std;

// unordered_map<char, int> icp{{'#, 0'}, {'(', 6}, {'*', 4}, {'/', 4}, {'%', 4},
//     {'+', 2}, {'-', 2}, {')', 1}};
//  isp;
unordered_map<char, int> icp, isp;
// 栈外(in coming priority)优先数、
// 栈内(in stack priority)优先数
int main() {
    // icp.insert(pair<char, int>('#', 0));
    // icp.insert(unordered_map<char, int>::value_type('#', 0));
    return 0;
}

int postfix(string expression) {
    stack<char> stk;
    char ch = '#', ch1, op;
    stk.push(ch);
    return 0;
}